cache.go

```diff
@@ -6,67 +6,124 @@ import (
 	"io"
 	"os"
 	"path/filepath"
+	"strconv"
+	"strings"
 	"sync"
+	"time"
 
 	openai "github.com/sashabaranov/go-openai"
 )
 
+// CacheType represents the type of cache being used.
+type CacheType string
+
+// Cache types for different purposes.
+const (
+	ConversationCache CacheType = "conversations"
+	TemporaryCache    CacheType = "temp"
+)
+
 const cacheExt = ".gob"
 
 var errInvalidID = errors.New("invalid id")
 
-type convoCache struct {
-	dir string
+// Cache is a generic cache implementation that stores data in files.
+type Cache[T any] struct {
+	baseDir string
+	cType   CacheType
 }
 
-func newCache(dir string) *convoCache {
-	return &convoCache{dir}
+// NewCache creates a new cache instance with the specified base directory and cache type.
+func NewCache[T any](baseDir string, cacheType CacheType) (*Cache[T], error) {
+	cacheDir := filepath.Join(baseDir, string(cacheType))
+	if err := os.MkdirAll(cacheDir, os.ModePerm); err != nil {
+		return nil, fmt.Errorf("create cache directory: %w", err)
+	}
+	return &Cache[T]{
+		baseDir: baseDir,
+		cType:   cacheType,
+	}, nil
 }
 
-func (c *convoCache) read(id string, messages *[]openai.ChatCompletionMessage) error {
+func (c *Cache[T]) cacheDir() string {
```

Comment:
```suggestion
func (c *Cache[T]) dir() string {
```

no nead to repeat cache I think

Reply:
done on main